### PR TITLE
Nullsafety changes

### PR DESCRIPTION
**Pull request by j4qfrost**

### CODE REVIEW

File: aqueduct/lib/src/db/postgresql/postgresql_query.dart [hunk 1]
```diff
@@ -43,11 +43,11 @@ class PostgresQuery<InstanceType extends ManagedObject> extends Object
 
     buffer.write("VALUES (${builder.sqlValuesToInsert}) ");
 
-    if ((builder.returning?.length ?? 0) > 0) {
+    if (builder.returning.isNotEmpty) {
```
Comment (j4qfrost) [inline aqueduct/lib/src/db/postgresql/postgresql_query.dart:46]: We are guaranteed that `returning` is not null. Cleaned up the check

File: aqueduct/lib/src/db/postgresql/query_builder.dart [hunk 2]
```diff
@@ -12,8 +12,10 @@ class PostgresQueryBuilder extends TableBuilder {
       : valueKeyPrefix = "v${prefixIndex}_",
         placeholderKeyPrefix = "@v${prefixIndex}_",
         super(query) {
-    (query.valueMap ?? query.values?.backing?.contents)
-        .forEach(addColumnValueBuilder);
+    final valueMap = query.valueMap ?? query.values.backing.contents;
```
Comment (j4qfrost) [inline aqueduct/lib/src/db/postgresql/query_builder.dart:15]: This is clearer to read and conforms with nnbd.

File: aqueduct/lib/src/db/postgresql/query_builder.dart [hunk 3]
```diff
@@ -33,33 +35,35 @@ class PostgresQueryBuilder extends TableBuilder {
...
-    return predicate.format;
+    return predicate?.format;
   }
 
   void addColumnValueBuilder(String key, dynamic value) {
     final builder = _createColumnValueBuilder(key, value);
-    columnValueBuildersByKey[builder.sqlColumnName()] = builder;
-    variables[builder.sqlColumnName(withPrefix: valueKeyPrefix)] =
-        builder.value;
+    if (builder != null) {
```
Comment (j4qfrost) [inline aqueduct/lib/src/db/postgresql/query_builder.dart:50]: Opted for noop here, also considered catching a null check error and printing a warning.

File: aqueduct/lib/src/db/query/query.dart [hunk 4]
```diff
@@ -198,30 +198,30 @@ abstract class Query<InstanceType extends ManagedObject> {
   /// This is a safety measure for update and delete queries to prevent accidentally updating or deleting every row.
   /// This flag defaults to false, meaning that if this query is either an update or a delete, but contains no predicate,
   /// it will fail. If a query is meant to update or delete every row on a table, you may set this to true to allow this query to proceed.
-  bool canModifyAllInstances;
+  late bool canModifyAllInstances;
```
Comment (j4qfrost) [inline aqueduct/lib/src/db/query/query.dart:201]: Double check me here. late can get dicey.

**Comment by j4qfrost** [comment]
@Reductions I have added comments to highlight some of the additions I made to make your bulk inserts work with nnbd or improve readability.

@bsutton I know the linting changes are a bit much, but I would rather walk through the changes than revert the linting, there are lots of null check changes anyway. I think doing searches for `try catch` and `late` would be a good way to find some of the faults that I made. I am also less confident on some of the code generation for the tooling.

**Comment by Hexer10** [comment]
I've had a look around the PR, and I've seen most of the changes are added `?` around the code, I'm not sure if it was done to make the transition to nnbd quicker or it is intended to stay, but in my opinion it would be more feasible to actually check where a certain variable is not used in any `null` context are make in nnbd, otherwhise we are losing all the benefits from migrating to nnbd in my opionion.

Again in that regard in most of the "model" classes use setters and getters are used, while imo, the code would be much more concise, and it is going to handle better the benefits of nnbd, using immutable classes, along with a copyWith method (maybe generated with freezed or similar), for example:

From:
```
class AuthCode {
  String? code;
  late String clientID;
  int? resourceOwnerIdentifier;
  DateTime? issueDate;
  DateTime? expirationDate;
  bool? hasBeenExchanged;
  List<AuthScope>? requestedScopes;

  bool get isExpired {
    return expirationDate == null ||
        expirationDate!.difference(DateTime.now().toUtc()).inSeconds <= 0;
  }
}
```

To:
```
@immutable
class AuthCode {
  const AuthCode(
      {required this.code,
      required this.clientID,
      required this.resourceOwnerIdentifier,
      required this.issueDate,
      required this.expirationDate,
      required this.hasBeenExchanged,
      required this.requestedScopes});

  final String code;
  final String clientID;
  final int resourceOwnerIdentifier;
  final DateTime issueDate;
  final DateTime expirationDate;
  final bool hasBeenExchanged;
  final List<AuthScope> requestedScopes;

  bool get isExpired {
    return expirationDate.difference(DateTime.now().toUtc()).inSeconds <= 0;
  }

  AuthCode copyWith({...}) => AuthCode(...);
}
```

I know this would take sometime but it could make a nice overhaul of the project.

**Comment by bsutton** [comment]
I agree you with  respect to minimising the use of ? but the discussion to
date had been to minimise changes so we can get a stable release out.

Can I suggest you raise an issue to track your suggestion the
getters/setters.

**Comment by j4qfrost** [comment]
Having trouble sleeping, I will address you quickly. You are correct in
that we want a quick port, and I fully intend to help in reducing ? after
we figure out where we stand. There are lots of moving parts particularly
in the code generation that break if we use non-nullables. I have stated
those concerns before. I figured that the naive approach was a best first
step.

**Comment by bsutton** [comment]
Can I suggest maybe a middle ground.

Removing significant no.s of ? is actually straight forward except in two
cases
1) use of the late keyword
2) dynamic maps (looking at you Json)

I've not used the code generators and I agree that breakages are likely.

So the middle ground could be that we do an aggressive sweep of the
dependencies but leave any code around the generators with the more nieve
approach.

This will likely necessitate breaking API changes at a later date but we
can probably live with that in order to get 1.0 out the door.

**Comment by j4qfrost** [comment]
@Reductions should we address this as well? I will make partitions of the PR.

**Comment by j4qfrost** [comment]
@bsutton had some more time to digest your proposal and I agree. We should actually be prioritizing the dependencies. If we can make stable changes to dart-runtime, then where we can make improvements in aqueduct will present itself.

I do have a caveat. The objective of this PR was always to "make it work", not to optimize for nnbd. I think we need to have a clean slate for that, and be patient with approaching a satisfactory release. Unless we need to do a complete reengineering of the codebase to take advantage of nnbd, I think we should open another PR for performance.

@Reductions made a suggestion to me that the PR should be split into various parts, so as to do incremental reviews of ~30 vs ~300 files. I think it's a good idea, and a problem I caused by lack of foresight. I can find some time later to separate the changes into chunks.

**Comment by Hexer10** [comment]
Thanks to all for the replies to my comment. 
I was just wondering if this PR was aiming to just migrate the project and "make it work", or you'd want also to make consistent changes, though the first solution is definitely better.

**Pull request closed without merging**